src/ripple/shamap/impl/SHAMapDelta.cpp

```diff
@@ -286,4 +290,81 @@ SHAMap::walkMap(std::vector<SHAMapMissingNode>& missingNodes, int maxMissing)
     }
 }
 
+void
+SHAMap::walkMapParallel(
+    std::vector<SHAMapMissingNode>& missingNodes,
+    int maxMissing) const
+{
+    if (!root_->isInner())  // root_ is only node, and we have it
+        return;
+
+    using StackEntry = std::shared_ptr<SHAMapInnerNode>;
+    std::array<std::shared_ptr<SHAMapTreeNode>, 16> topChildren;
+    {
+        auto const& innerRoot =
+            std::static_pointer_cast<SHAMapInnerNode>(root_);
+        for (int i = 0; i < 16; ++i)
+        {
+            if (!innerRoot->isEmptyBranch(i))
+                topChildren[i] = descendNoStore(innerRoot, i);
+        }
+    }
+    std::vector<std::thread> workers;
+    workers.reserve(16);
+
+    std::array<std::stack<StackEntry, std::vector<StackEntry>>, 16> nodeStacks;
+
+    for (int rootChildIndex = 0; rootChildIndex < 16; ++rootChildIndex)
+    {
+        auto const& child = topChildren[rootChildIndex];
+        if (!child || !child->isInner())
+            continue;
+
+        nodeStacks[rootChildIndex].push(
+            std::static_pointer_cast<SHAMapInnerNode>(child));
+
+        JLOG(journal_.debug()) << "starting worker " << rootChildIndex;
+        std::mutex m;
```

Comment:
This `mutex` needs to be in the outermost scope.  We need `m`'s destructor not to run until after all the calls to `join()` have completed.